File: desktop/src/main/chatMemoryExtractor.ts
import { URL } from "node:url";

import type { ChatKnowledgeWrite } from "./chatPlan";
import type { MemoryCandidate } from "./memoryModel";

function unique<T>(values: T[]): T[] {
  return [...new Set(values)];
}

function normalizeWhitespace(value: string): string {
  return value.replace(/\s+/g, " ").trim();
}

function extractUrls(text: string): string[] {
  return unique(
    (text.match(/https?:\/\/[^\s)\]]+/g) ?? []).map((value) => value.replace(/[.,;]+$/, ""))
  );
}

function extractFullName(text: string): string | null {
  const match = text.match(
    /меня\s+зовут\s+([А-ЯЁA-Z][а-яёa-z-]+(?:\s+[А-ЯЁA-Z][а-яёa-z-]+){1,2})/i
  );
  return match ? normalizeWhitespace(match[1]) : null;
}

function extractOccupation(text: string): string | null {
  const match = text.match(
    /(?:^|[\s,.:;!?()])(?:я\s+)?(программист|разработчик|python[-\s]?разработчик|backend[-\s]?разработчик|frontend[-\s]?разработчик)(?=$|[\s,.:;!?()])/i
  );

  return match ? normalizeWhitespace(match[1].toLowerCase()) : null;
}

function extractStack(text: string): string | null {
  const stack: string[] = [];

  if (/\bpython\b/i.test(text)) {
    stack.push("Python");
  }

  if (/\bfastapi\b/i.test(text)) {
    stack.push("FastAPI");
  }

  if (/\breact\b/i.test(text)) {
    stack.push("React");
  }

  if (/\btypescript\b/i.test(text)) {
    stack.push("TypeScript");
  }

  return stack.length > 0 ? stack.join(", ") : null;
}

function extractInterest(text: string): string | null {
  if (
    /(нравится|люблю)\s+(?:изучать\s+)?нейросети/i.test(text) ||
    /\bнейросет/i.test(text)
  ) {
    return "Нейросети";
  }

  return null;
}

function extractQuietPreference(text: string): string | null {
  if (/(предпочитаю|люблю).{0,20}(тишин|в тишине)/i.test(text)) {
    return "Тишина";
  }

  return null;
}

function extractGpu(text: string): string | null {
  const match = text.match(
    /\b((?:AMD\s+Radeon|NVIDIA\s+GeForce|Intel\s+Arc)\s+[A-Za-z0-9!()+\-.\s]+?)(?=,|\.|;|$)/i
  );
  return match ? normalizeWhitespace(match[1]) : null;
}

function extractCpu(text: string): string | null {
  const labeledMatch = text.match(/(?:процессор|cpu)\s+([A-Za-z0-9+\-.\s]+?)(?=,|\.|;|$)/i);
  if (labeledMatch) {
    return normalizeWhitespace(labeledMatch[1]);
  }

  const directMatch = text.match(
    /\b((?:Ryzen|Intel\s+Core|Intel\s+Xeon|Apple\s+M)\s+[A-Za-z0-9+\-.\s]+?)(?=,|\.|;|$)/i
  );
  return directMatch ? normalizeWhitespace(directMatch[1]) : null;
}

function normalizeCapacity(value: string, unit: string): string {
  return `${value} ${unit.toUpperCase()}`;
}

function extractRam(text: string): string | null {
  const match = text.match(
    /(\d+)\s*(gb|гб|tb|тб)\s*(?:ozu|озу|ram|оператив(?:ной)? памяти?)/i
  );
  return match ? normalizeCapacity(match[1], /tb|тб/i.test(match[2]) ? "TB" : "GB") : null;
}

function extractStorage(text: string): string | null {
  const match = text.match(
    /(?:диск|ssd|hdd|накопитель)(?:\s+\w+)*\s+(?:на\s*)?(\d+)\s*(gb|гб|tb|тб)/i
  );
  return match ? normalizeCapacity(match[1], /tb|тб/i.test(match[2]) ? "TB" : "GB") : null;
}

function extractObservation(text: string): MemoryCandidate[] {
  const observations: MemoryCandidate[] = [];

  if (/(грустн|тяжело|тревожн|уста(л|ю|ет|ем))/i.test(text)) {
    observations.push({
      kind: "observation",
      target: "assist/observations",
      key: "recent_emotional_signal",
      value: "Пользователь описывает устойчиво негативное или уставшее состояние.",
      confidence: "inferred"
    });
  }

  if (/(ошибк|опечатк)/i.test(text) && /(пишу|пишет|грамот)/i.test(text)) {
    observations.push({
      kind: "observation",
      target: "assist/observations",
      key: "communication_style",
      value: "Возможны устойчивые орфографические ошибки, стоит упрощать формулировки.",
      confidence: "inferred"
    });
  }

  return observations;
}

export function extractChatMemoryCandidates(text: string): MemoryCandidate[] {
  const candidates: MemoryCandidate[] = [];
  const normalized = normalizeWhitespace(text);

  const fullName = extractFullName(normalized);
  if (fullName) {
    candidates.push({
      kind: "fact",
      target: "assist/profile",
      key: "full_name",
      value: fullName,
      confidence: "direct"
    });
  }

  const occupation = extractOccupation(normalized);
  if (occupation) {
    candidates.push({
      kind: "fact",
      target: "assist/profile",
      key: "occupation",
      value: occupation,
      confidence: "direct"
    });
  }

  const preferredStack = extractStack(normalized);
  if (preferredStack) {
    candidates.push({
      kind: "preference",
      target: "assist/preferences",
      key: "preferred_stack",
      value: preferredStack,
      confidence: "direct"
    });
  }

  const interest = extractInterest(normalized);
  if (interest) {
    candidates.push({
      kind: "preference",
      target: "assist/preferences",
      key: "interests",
      value: interest,
      confidence: "direct"
    });
  }

  const quietPreference = extractQuietPreference(normalized);
  if (quietPreference) {
    candidates.push({
      kind: "preference",
      target: "assist/preferences",
      key: "preferred_environment",
      value: quietPreference,
      confidence: "direct"
    });
  }

  const gpu = extractGpu(normalized);
  if (gpu) {
    candidates.push({
      kind: "fact",
      target: "assist/profile",
      key: "gpu",
      value: gpu,
      confidence: "direct"
    });
  }

  const cpu = extractCpu(normalized);
  if (cpu) {
    candidates.push({
      kind: "fact",
      target: "assist/profile",
      key: "cpu",
      value: cpu,
      confidence: "direct"
    });
  }

  const ram = extractRam(normalized);
  if (ram) {
    candidates.push({
      kind: "fact",
      target: "assist/profile",
      key: "ram",
      value: ram,
      confidence: "direct"
    });
  }

  const storage = extractStorage(normalized);
  if (storage) {
    candidates.push({
      kind: "fact",
      target: "assist/profile",
      key: "storage",
      value: storage,
      confidence: "direct"
    });
  }

  candidates.push(...extractObservation(normalized));

  for (const sourceUrl of extractUrls(normalized)) {
    try {
      const parsed = new URL(sourceUrl);
      candidates.push({
        kind: "source",
        target: "assist/docs/websites",
        key: `${parsed.protocol}//${parsed.host}`,
        value: parsed.host,
        confidence: "direct"
      });
      candidates.push({
        kind: "source",
        target: "assist/docs/papers",
        key: sourceUrl,
        value: sourceUrl,
        confidence: "direct"
      });
    } catch {
      continue;
    }
  }

  return candidates;
}

export function extractChatMemoryWrites(text: string): ChatKnowledgeWrite[] {
  return extractChatMemoryCandidates(text).map((candidate) => ({
    target: candidate.target,
    key: candidate.key,
    value: candidate.value
  }));
}

File: bot/app/chat_memory_extractor.py
from __future__ import annotations

import re
from dataclasses import asdict, dataclass


URL_PATTERN = re.compile(r"https?://[^\s)\]]+")
FULL_NAME_PATTERN = re.compile(
    r"\bменя\s+зовут\s+([А-ЯЁ][а-яё]+(?:\s+[А-ЯЁ][а-яё]+){1,2})",
    re.IGNORECASE,
)
ROLE_PATTERN = re.compile(
    r"(?:^|[\s,.:;!?()])(?:я\s+)?(программист|разработчик|python[-\s]?разработчик|backend[-\s]?разработчик|frontend[-\s]?разработчик)(?=$|[\s,.:;!?()])",
    re.IGNORECASE,
)
GPU_PATTERN = re.compile(
    r"\b((?:AMD\s+Radeon|NVIDIA\s+GeForce|Intel\s+Arc)\s+[A-Za-z0-9!()+\-.\s]+?)(?=,|\.|;|$)",
    re.IGNORECASE,
)
CPU_LABELED_PATTERN = re.compile(
    r"(?:процессор|cpu)\s+([A-Za-z0-9+\-.\s]+?)(?=,|\.|;|$)",
    re.IGNORECASE,
)
CPU_DIRECT_PATTERN = re.compile(
    r"\b((?:Ryzen|Intel\s+Core|Intel\s+Xeon|Apple\s+M)\s+[A-Za-z0-9+\-.\s]+?)(?=,|\.|;|$)",
    re.IGNORECASE,
)
RAM_PATTERN = re.compile(
    r"(\d+)\s*(gb|гб|tb|тб)\s*(?:ozu|озу|ram|оператив(?:ной)? памяти?)",
    re.IGNORECASE,
)
STORAGE_PATTERN = re.compile(
    r"(?:диск|ssd|hdd|накопитель)(?:\s+\w+)*\s+(?:на\s*)?(\d+)\s*(gb|гб|tb|тб)",
    re.IGNORECASE,
)

STACK_LABELS: tuple[tuple[str, str], ...] = (
    ("fastapi", "FastAPI"),
    ("python", "Python"),
    ("typescript", "TypeScript"),
    ("react", "React"),
    ("mcp", "MCP"),
    ("docker", "Docker"),
    ("postgres", "PostgreSQL"),
)


@dataclass(frozen=True, slots=True)
class ConversationMemoryWrite:
    target: str
    key: str
    value: str


def _normalize_space(value: str) -> str:
    return " ".join(value.split()).strip()


def _normalize_capacity(value: str, unit: str) -> str:
    return f"{value} {'TB' if unit.casefold() in {'tb', 'тб'} else 'GB'}"


def extract_memory_writes(text: str) -> list[ConversationMemoryWrite]:
    writes: list[ConversationMemoryWrite] = []
    normalized = _normalize_space(text)

    full_name_match = FULL_NAME_PATTERN.search(normalized)
    if full_name_match is not None:
        writes.append(
            ConversationMemoryWrite(
                target="assist/profile",
                key="full_name",
                value=_normalize_space(full_name_match.group(1)),
            )
        )

    role_match = ROLE_PATTERN.search(normalized)
    if role_match is not None:
        writes.append(
            ConversationMemoryWrite(
                target="assist/profile",
                key="occupation",
                value=_normalize_space(role_match.group(1).lower()),
            )
        )

    detected_stack = [
        label for needle, label in STACK_LABELS if needle in normalized.casefold()
    ]
    if detected_stack:
        writes.append(
            ConversationMemoryWrite(
                target="assist/preferences",
                key="preferred_stack",
                value=", ".join(dict.fromkeys(detected_stack)),
            )
        )

    if re.search(r"(нравится|люблю)\s+(?:изучать\s+)?нейросети", normalized, re.IGNORECASE) or re.search(
        r"\bнейросет", normalized, re.IGNORECASE
    ):
        writes.append(
            ConversationMemoryWrite(
                target="assist/preferences",
                key="interests",
                value="Нейросети",
            )
        )

    gpu_match = GPU_PATTERN.search(normalized)
    if gpu_match is not None:
        writes.append(
            ConversationMemoryWrite(
                target="assist/profile",
                key="gpu",
                value=_normalize_space(gpu_match.group(1)),
            )
        )

    cpu_match = CPU_LABELED_PATTERN.search(normalized) or CPU_DIRECT_PATTERN.search(normalized)
    if cpu_match is not None:
        writes.append(
            ConversationMemoryWrite(
                target="assist/profile",
                key="cpu",
                value=_normalize_space(cpu_match.group(1)),
            )
        )

    ram_match = RAM_PATTERN.search(normalized)
    if ram_match is not None:
        writes.append(
            ConversationMemoryWrite(
                target="assist/profile",
                key="ram",
                value=_normalize_capacity(ram_match.group(1), ram_match.group(2)),
            )
        )

    storage_match = STORAGE_PATTERN.search(normalized)
    if storage_match is not None:
        writes.append(
            ConversationMemoryWrite(
                target="assist/profile",
                key="storage",
                value=_normalize_capacity(storage_match.group(1), storage_match.group(2)),
            )
        )

    for url in extract_source_urls(text):
        host_match = re.match(r"https?://([^/]+)", url)
        host = host_match.group(1) if host_match is not None else url
        writes.append(
            ConversationMemoryWrite(
                target="assist/docs/websites",
                key=f"https://{host}",
                value=host,
            )
        )
        writes.append(
            ConversationMemoryWrite(
                target="assist/docs/papers",
                key=url,
                value=url,
            )
        )

    return writes


def extract_source_urls(*chunks: str) -> list[str]:
    urls: list[str] = []
    for chunk in chunks:
        for match in URL_PATTERN.findall(chunk):
            normalized = match.rstrip(".,;")
            if normalized not in urls:
                urls.append(normalized)
    return urls


def serialize_memory_writes(writes: list[ConversationMemoryWrite]) -> list[dict[str, str]]:
    return [asdict(write) for write in writes]
